File: C/LeetCode/003-LongestSubstringWithoutRepeatingCharacters.c
#include "003-LongestSubstringWithoutRepeatingCharacters.h"

int lengthOfLongestSubstring(char *s)
{
	if (s == NULL) { return 0; }

	int map[256];
	int i;
	int maxLen;
	int lastRepeatPos;

	for (i = 0; i < 256; i++) {
		map[i] = -1;
	}

	maxLen = 0;
	lastRepeatPos = -1;
	for (i = 0; s[i] != '\0'; i++)
	{
		if (map[s[i]] != -1 && lastRepeatPos < map[s[i]])
		{
			lastRepeatPos = map[s[i]];
		}
		if (maxLen < i - lastRepeatPos)
		{
			maxLen = i - lastRepeatPos;
		}
		map[s[i]] = i;
	}

	return maxLen;
}

File: Java/src/main/java/bigegg/leetcode/_003_LongestSubstringWithoutRepeatingCharacters.java
package bigegg.leetcode;

import java.util.HashMap;
import java.util.Map;

public class _003_LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }

        Map<Character, Integer> map = new HashMap<Character, Integer>();
        int maxLen = 0;
        int lastRepeatPos = -1;
        for (int i = 0; i < s.length(); i++) {
            if (map.containsKey(s.charAt(i)) && lastRepeatPos < map.get(s.charAt(i))) {
                lastRepeatPos = map.get(s.charAt(i));
            }
            if (maxLen < i - lastRepeatPos) {
                maxLen = i - lastRepeatPos;
            }
            map.put(s.charAt(i), i);
        }

        return maxLen;
    }
}
